src/cargo/core/package.rs

```diff
@@ -571,6 +558,66 @@ impl<'cfg> PackageSet<'cfg> {
         Ok(())
     }
 
+    /// Check if there are any dependency packages that do not have any libs.
+    pub(crate) fn no_lib_pkgs(
+        &self,
+        resolve: &Resolve,
+        root_ids: &[PackageId],
+        has_dev_units: HasDevUnits,
+        requested_kinds: &[CompileKind],
+        target_data: &RustcTargetData<'_>,
+        force_all_targets: ForceAllTargets,
+    ) -> CargoResult<Vec<&Package>> {
+        let mut ret = vec![];
+
+        root_ids.iter().for_each(|pkg_id| {
+            PackageSet::filter_deps(
+                *pkg_id,
+                resolve,
+                has_dev_units,
+                requested_kinds,
+                target_data,
+                force_all_targets,
+            )
+            .for_each(|(package_id, _)| {
+                if let Ok(dep_pkg) = self.get_one(package_id) {
+                    if !dep_pkg.targets().iter().any(|t| t.is_lib()) {
+                        ret.push(dep_pkg);
+                    }
+                }
+            });
+        });
+
+        Ok(ret)
+    }
+
+    fn filter_deps<'a>(
+        pkg_id: PackageId,
+        resolve: &'a Resolve,
+        has_dev_units: HasDevUnits,
+        requested_kinds: &'a [CompileKind],
+        target_data: &'a RustcTargetData<'_>,
+        force_all_targets: ForceAllTargets,
+    ) -> impl Iterator<Item = (PackageId, &'a HashSet<Dependency>)> {
```

Comment:
afaict - both callsites to this don't actually use the HashSet<Dependency> in the result. Would it make sense to simplify with some kind of `.map(|(a, b)| a)` at the end?

Reply:
Good catch! Thank you!